File: hosts/jetbrains/src/main/kotlin/dev/oxp/jetbrains/edh/EdhMarker.kt
package dev.oxp.jetbrains.edh

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths

/**
 * Cross-host EDH (Extension Development Host) handshake marker.
 *
 * `oxp dev` writes a JSON marker at `$OXP_HOME/edh/autostart.json`
 * (default `~/.oxp/edh/autostart.json`) after binding its WebSocket
 * port. The plugin's [EdhStartupActivity] reads it on every project
 * open, validates folder + freshness, then attaches to `wsUrl`
 * directly — there is exactly one `oxp dev` per session.
 *
 * Schema is v1, identical to `packages/cli/src/lib/edh-marker.ts`:
 *
 *   { v:1, ts, folderPath, wsUrl, forkBin? }
 *
 * Unknown future fields are ignored.
 */
object EdhMarker {
    private val json = Json { ignoreUnknownKeys = true }
    private const val TTL_MS = 60_000L
    const val SCHEMA_VERSION = 1

    @Serializable
    data class Payload(
        val v: Int = SCHEMA_VERSION,
        val ts: Long,
        val folderPath: String,
        val wsUrl: String = "",
        val forkBin: String? = null,
    )

    fun path(): Path {
        val home = System.getenv("OXP_HOME")
            ?: Paths.get(
                System.getProperty("user.home") ?: "/tmp",
                ".oxp",
            ).toString()
        return Paths.get(home, "edh", "autostart.json")
    }

    /**
     * In-process helper used by [StartDevSessionAction]. Writes a marker
     * with the current timestamp and folder path. No wsUrl is set: the
     * consuming window will spawn `oxp dev` itself in that legacy flow.
     */
    fun write(folderPath: String) {
        val p = path()
        Files.createDirectories(p.parent)
        val payload = Payload(
            ts = System.currentTimeMillis(),
            folderPath = folderPath,
        )
        Files.writeString(p, json.encodeToString(Payload.serializer(), payload))
    }

    /**
     * Consume the marker iff it matches `projectBasePath` and is < 60s
     * old. Returns the parsed payload (so the caller can read `wsUrl`)
     * or null if no fresh, matching marker is present. The marker is
     * deleted on consume so concurrent windows don't double-fire.
     */
    fun consumeIfMatches(projectBasePath: String?): Payload? {
        if (projectBasePath == null) return null
        val p = path()
        if (!Files.exists(p)) return null

        val payload: Payload = try {
            json.decodeFromString(Payload.serializer(), Files.readString(p))
        } catch (_: Exception) {
            tryDelete(p)
            return null
        }

        if (System.currentTimeMillis() - payload.ts > TTL_MS) {
            tryDelete(p)
            return null
        }

        if (!sameFolder(payload.folderPath, projectBasePath)) {
            // Not for us — leave the marker for the right window.
            return null
        }

        tryDelete(p)
        return payload
    }

    /**
     * Compare two folder paths for "same target" without forcing the
     * caller to canonicalize. Handles trailing slashes and macOS's
     * `/private/var` ↔ `/var` symlink quirk where realpath differs.
     */
    private fun sameFolder(a: String, b: String): Boolean {
        if (a == b) return true
        val na = a.trimEnd('/', '\\')
        val nb = b.trimEnd('/', '\\')
        if (na == nb) return true
        return try {
            Paths.get(na).toRealPath() == Paths.get(nb).toRealPath()
        } catch (_: Exception) {
            false
        }
    }

    private fun tryDelete(p: Path) {
        try {
            Files.deleteIfExists(p)
        } catch (_: Exception) {
            // best-effort
        }
    }
}

File: packages/cli/src/lib/edh-marker.ts
/**
 * Cross-host EDH (Extension Development Host) handshake marker.
 *
 * `oxp dev` writes a JSON marker after it has bound a port. The host
 * extension running in the freshly-spawned IDE window reads it on
 * activate, validates folder + freshness, then connects to `wsUrl`
 * directly. The host never spawns its own CLI — there is exactly one
 * `oxp dev` per session, which avoids port collisions on every fork
 * (VS Code, Cursor, Windsurf, Antigravity, VSCodium, …).
 *
 * Schema is versioned so future fields don't silently break old hosts.
 */
import { existsSync, mkdirSync, unlinkSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";

export const EDH_MARKER_VERSION = 1;

export interface EdhMarker {
  v: number;
  ts: number;
  folderPath: string;
  /** Full WebSocket URL the host should connect to (e.g. ws://127.0.0.1:7373/dev). */
  wsUrl: string;
  /** CLI binary name of the fork that spawned the window (debug/diagnostics). */
  forkBin?: string;
}

export function edhMarkerPath(): string {
  const home =
    process.env.OXP_HOME ??
    join(process.env.HOME ?? process.env.USERPROFILE ?? "/tmp", ".oxp");
  return join(home, "edh", "autostart.json");
}

export function writeEdhMarker(marker: Omit<EdhMarker, "v" | "ts">): string {
  const full: EdhMarker = {
    v: EDH_MARKER_VERSION,
    ts: Date.now(),
    ...marker,
  };
  const p = edhMarkerPath();
  mkdirSync(dirname(p), { recursive: true });
  writeFileSync(p, JSON.stringify(full), "utf8");
  return p;
}

export function deleteEdhMarker(): void {
  const p = edhMarkerPath();
  if (!existsSync(p)) return;
  try {
    unlinkSync(p);
  } catch {
    /* noop */
  }
}
